### PR TITLE
[Draft] Migrate the CLI to v2 of the golang sdk

### PR DESCRIPTION
**Pull request by wilsonge**
Description of changes:

- Initial work on migrating the app to version 2 of the SDK (App has largely been migrated to the new SDK's with equivalent functionality). Obviously this would be a new major version of the product.

## Current list of items to finish
- [x] Some functionality has been dropped relating to the injection of regions from different sources which needs restoring (cmd/main.go)
- [ ] There's still a hard dependency on the v1 sdk endpoints package in odpricing.go which will likely require some work on the smithy side of things to generate a mapping between regions and descriptions (it's in the smithy generated json file but not in the golang sdk).
- [x] Tests are still a WIP and some more minor restructuring will likely be required as part of that.
- [x] Propagation of contexts through the stack

## Why Migrate to v2
- ~At the end of this the overall binary should be a bit smaller by the modular nature of the new sdk~ i was wrong it's bigger - what can you do 🤷 
- Some projects that depend on this (e.g. eksctl) are currently having to create two sessions - one in the v2 sdk for all their direct dependencies and one in the v1 sdk for this project
- Stronger typehints on the vm wrapper

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by bwagner5** [comment]
Hey! Thanks for working on this! Curious what the binary size diff was on the WIP for this PR? Seems valuable to do anyways to support clients using v2 versus v1. I suppose if a client is using v1 though, this change would pull in the additional v2 deps for their binary (I think kOps uses v1). I would guess the binary sizes wouldn't be radically different though since the unused packages would shaken out during compilation.

**Comment by wilsonge** [comment]
Interestingly bigger. Looks like the ec2 serializers are pretty hefty https://github.com/aws/aws-sdk-go-v2/blob/main/service/ec2/serializers.go about 60k lines :) about 33MB for now - but 3MB of that is the v1 package endpoints - which like I say can probably be removed in favour of smithy generated stuff.

I've committed fixes for the last 2 tests (at least they pass now locally). Would be good to get a retrigger on the tests here to see if they pass on the git workflows

**Comment by snay2** [comment]
@wilsonge We've been maintaining the `THIRD_PARTY_LICENSES` file by hand. You can copy/paste and follow the same formatting for any new libraries.

**Comment by wilsonge** [comment]
@snay2 / @bwagner5 sorry to be a pain but could i get you to retrigger the workflow please

**Comment by wilsonge** [comment]
OK So tests here look good. What's the next steps. I think this one needs some proper human code review too because there's an awful lot of changes.

**Comment by snay2** [comment]
@wilsonge Have you finished the tasks you outlined in the PR description? If so, I'd say check those off and remove "draft" from the title, and we'll start reviewing the code.

**Review by cjerad — CHANGES_REQUESTED** [review]
Hi wilsonge, in addition to the outstanding tasks I have a few questions and suggestions.

**Comment by TiberiuGC** [comment]
Hi 👋 - reaching out on eksctl behalf, we're also looking forward to completely migrate to go sdk v2. Is there any chance the work on this PR will be picked up from where it was left in the foreseeable future?

**Comment by jillmon** [comment]
hi @TiberiuGC. I wasn't aware that eksctl had a dependency on EC2 Instance Selector. Can you give us more detail as to how you're utilizing this tool. It would help with our prioritization of this PR because it looks like this may require another developer to pick up where @wilsonge has left off.

**Comment by TiberiuGC** [comment]
@jillmon when creating (clusters with) nodegroups using eksctl, users have the option to specify the properties they want for their nodes e.g.
```
eksctl create nodegroup --instance-selector-
--instance-selector-cpu-architecture  -- x86_64, or arm64
--instance-selector-gpus              -- an integer value
--instance-selector-memory            -- 4 or 4GiB
--instance-selector-vcpus             -- an integer value (2, 4 etc)
```
and we let instance selector do it's magic on deciding the instance types.

As to why this PR is relevant for us, I think @wilsonge explained it well in the description

> **Why Migrate to V2**
...
Some projects that depend on this (e.g. eksctl) are currently having to create two sessions - one in the v2 sdk for all their direct dependencies and one in the v1 sdk for this project. 

Let me know if I can provide any more info 🙏

File: pkg/ec2pricing/ec2pricing.go [hunk 1]
```diff
@@ -40,31 +41,49 @@ type EC2Pricing struct {
...
 
 // New creates an instance of instance-selector EC2Pricing
-func New(sess *session.Session) *EC2Pricing {
+func New(ctx context.Context) *EC2Pricing {
 	// use us-east-1 since pricing only has endpoints in us-east-1 and ap-south-1
-	pricingClient := pricing.New(sess.Copy(aws.NewConfig().WithRegion("us-east-1")))
+	cfg, err := config.LoadDefaultConfig(ctx, config.WithRegion("us-east-1"))
+	if err != nil {
+		panic("configuration error, " + err.Error())
+	}
```
[the remark this answers is not in the document]
Comment (wilsonge) [reply pkg/ec2pricing/ec2pricing.go]: Done but worth noting the function currently doesn't return an error so makes the b/c impact larger

File: pkg/awsapi/instancetypesec2.go [hunk 2]
```diff
@@ -0,0 +1,7 @@
+package awsapi
+
+import "github.com/aws/aws-sdk-go-v2/service/ec2"
+
+type InstanceTypesEC2Interface interface {
+	ec2.DescribeInstanceTypesAPIClient
+}
```
[the remark this answers is not in the document]
Comment (wilsonge) [reply pkg/awsapi/instancetypesec2.go]: No clue :D I added them in for testing at some point but no clue why anymore. Removed again :)

File: cmd/main.go [hunk 3]
```diff
@@ -289,48 +291,84 @@ Full docs can be found at github.com/aws/amazon-` + binName
...
 	filters := selector.Filters{
-		VCpusRange:                       cli.IntRangeMe(flags[vcpus]),
+		VCpusRange:                       cli.Int32RangeMe(flags[vcpus]),
 		MemoryRange:                      cli.ByteQuantityRangeMe(flags[memory]),
 		VCpusToMemoryRatio:               cli.Float64Me(flags[vcpusToMemoryRatio]),
-		CPUArchitecture:                  cli.StringMe(flags[cpuArchitecture]),
-		CPUManufacturer:                  cli.StringMe(flags[cpuManufacturer]),
-		GpusRange:                        cli.IntRangeMe(flags[gpus]),
+		CPUArchitecture:                  &cpuArchitectureFilterValue,
+		CPUManufacturer:                  &cpuManufacturerFilterValue,
```
[the remark this answers is not in the document]
Comment (wilsonge) [reply cmd/main.go:356]: Correct but my assumption (rightly or wrongly) was that this sort of stuff would bubble up in the integration tests you guys have. I don't think there's any easy workaround - it's a fallout of the stricter type definitions in the new sdk.

**Comment by wilsonge** [comment]
I've rebased this onto the latest main branch and applied most the review comments. There's still my todo's listed above (on me but if someone else wants to pick them up as I don't have huge amounts of time just let me know) and the outstanding item about the region lists which I still believe needs an AWS staff member to be resolved.

File: pkg/ec2pricing/odpricing.go [hunk 4]
```diff
@@ -157,7 +195,7 @@ func (c *OnDemandPricing) Save() error {
 	if err != nil {
 		return err
 	}
-	os.Mkdir(c.DirectoryPath, 0755)
+	_ = os.Mkdir(c.DirectoryPath, 0755)
```
Comment (cjerad) [inline pkg/ec2pricing/odpricing.go:198]: Please handle error

File: pkg/ec2pricing/odpricing.go [hunk 5]
```diff
@@ -68,11 +106,11 @@ func LoadODCacheOrNew(pricingClient pricingiface.PricingAPI, region string, full
 		cache:          cache.New(fullRefreshTTL, fullRefreshTTL),
 	}
 	if fullRefreshTTL <= 0 {
-		odPricing.Clear()
+		_ = odPricing.Clear()
```
Comment (cjerad) [inline pkg/ec2pricing/odpricing.go:109]: Please handle error

**Review by cjerad — CHANGES_REQUESTED** [review]
Hi wilsonge, thanks for updating this. Just a few small suggestions.

File: cmd/main.go [hunk 6]
```diff
@@ -246,14 +246,19 @@ Full docs can be found at github.com/aws/amazon-` + binName
 		log.Println("--service eks is deprecated. EKS generally supports all instance types")
 	}
 
-	sess, err := getRegionAndProfileAWSSession(cli.StringMe(flags[region]), cli.StringMe(flags[profile]))
+	ctx := context.Background()
```
[the remark this answers is not in the document]
Comment (wilsonge) [reply cmd/main.go:249]: This is to allow us to remove the registerShutdown line in this file in a more native way? Just for my understanding

File: pkg/selector/selector.go [hunk 7]
```diff
@@ -182,15 +194,17 @@ func (itf Selector) truncateResults(maxResults *int, instanceTypeInfoSlice []*in
...
 		TransformFn(itf.TransformBaseInstanceType),
 		TransformFn(itf.TransformFlexible),
 		TransformFn(itf.TransformForService),
 	}
 	var err error
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
 	for _, transform := range transforms {
-		filters, err = transform.Transform(filters)
+		filters, err = transform.Transform(ctx, filters)
```
[the remark this answers is not in the document]
Comment (wilsonge) [reply pkg/selector/selector.go]: Not sure now. Guess it wasn't :)

File: cmd/main.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (wilsonge) [reply cmd/main.go]: Implemented :)

**Comment by wilsonge** [comment]
Pushed up the latest changes requested. I've also pushed a change to better handle the aws context usage for the pricing client that will need a quick review on top of the previous stuff.

**Review by cjerad — CHANGES_REQUESTED** [review]

**Comment by cjerad** [comment]
Resolved merge conflicts locally then squashed and pushed all changes.

**Pull request closed without merging**